source/common/upstream/cluster_factory_impl.cc

```diff
@@ -96,8 +96,17 @@ ClusterFactoryImplBase::selectDnsResolver(const envoy::config::cluster::v3::Clus
     for (const auto& resolver_addr : resolver_addrs) {
       resolvers.push_back(Network::Address::resolveProtoAddress(resolver_addr));
     }
-    const bool use_tcp_for_dns_lookups = cluster.use_tcp_for_dns_lookups();
-    return context.dispatcher().createDnsResolver(resolvers, use_tcp_for_dns_lookups);
+
+    auto dns_resolver_options =
+        envoy::config::core::v3::DnsResolverOptions(cluster.dns_resolver_options());
+    // Field bool `use_tcp_for_dns_lookups` will be deprecated in future. To keep supporting earlier
+    // implementation of control plane APIs only accept this value if `use_tcp_for_dns_lookups`
+    // is not set via dns_resolver_options but is set `true` in the field bool
+    // `use_tcp_for_dns_lookups`.
+    if (cluster.use_tcp_for_dns_lookups() && !dns_resolver_options.has_use_tcp_for_dns_lookups()) {
+      dns_resolver_options.mutable_use_tcp_for_dns_lookups()->set_value(true);
+    }
+    return context.dispatcher().createDnsResolver(resolvers, dns_resolver_options);
```

Comment:
See the comment above about eliminating the `has_XXXX` functions.

You could declare a new `dns_resolver_options` and set  `use_tcp_for_dns_lookups =  cluster.use_tcp_for_dns_lookups()` instead of a copy and modify.

As it is written here if `has_use_tcp_for_dns_lookups()` returns false, then we'd never execute line 107 since both conditions on 106 collectively evaluate to false.

Reply:
The idea here is to still support _use_tcp_for_dns_lookups_ even though it is marked deprecated.

So the fields here are:
1. _use_tcp_for_dns_lookups_ <-- deprecated, but can still be programmed by old xDS control servers
2. _use_tcp_for_dns_lookups_  **via** _dns_resolver_options_
3. _no_default_search_domain_ **via** _dns_resolver_options_

In the Line 106 with "**if condition**", what we are checking is if _dns_resolver_options_ is configured or not (possibly old implementation of xDS controller will not configure this value).

And in the "**if condition**" if the field _dns_resolver_options.use_tcp_for_dns_lookups_ is not programmed but the field _use_tcp_for_dns_lookups_ (deprecated) is programmed then we will update the _use_tcp_for_dns_lookups_ field inside variable _dns_resolver_options_ as true.


Also to be noted that in this code the preference is given for _config.dns_resolver_options.use_tcp_for_dns_lookups_ over bare boolean field _config.use_tcp_for_dns_lookups_.

Reply:
made this logic simpler, as part of the commit to change type of use_tcp_for_dns_lookups & no_default_search_domain from BoolValue to simply bool.